File: scripts/voice/tts.py
#!/usr/bin/env python3
"""
Local TTS (Text-to-Speech) using Kokoro or Windows SAPI fallback.
Usage: python tts.py <text> [--voice VOICE] [--output OUTPUT]
       python tts.py --file <text_file> [--voice VOICE] [--output OUTPUT]
Returns: JSON with audio file path
"""
import argparse
import json
import os
import sys
import tempfile
import wave


def tts_kokoro(text: str, voice: str = "af_bella", output_path: str = None) -> dict:
    """Synthesize speech using Kokoro TTS."""
    try:
        from kokoro import KPipeline
        import soundfile as sf
    except ImportError:
        return {"error": "kokoro not installed. Run: pip install kokoro", "text": ""}

    if not output_path:
        output_path = os.path.join(tempfile.gettempdir(), f"kokoro_tts_{abs(hash(text))}.wav")

    try:
        pipeline = KPipeline(lang_code=voice[0])  # 'a' for American English, 'b' for British, etc.
        generator = pipeline(text, voice=voice, speed=1.0)

        all_audio = []
        for i, (graphemes, phonemes, audio) in enumerate(generator):
            all_audio.append(audio)

        if all_audio:
            import numpy as np
            combined = np.concatenate(all_audio)
            sf.write(output_path, combined, 24000)
            return {"audio_path": output_path, "format": "wav", "engine": "kokoro"}
        else:
            return {"error": "No audio generated", "text": ""}
    except Exception as e:
        return {"error": f"Kokoro TTS failed: {str(e)}", "text": ""}


def tts_sapi_windows(text: str, output_path: str = None) -> dict:
    """Synthesize speech using Windows SAPI (built-in, no deps needed)."""
    if not output_path:
        output_path = os.path.join(tempfile.gettempdir(), f"sapi_tts_{abs(hash(text))}.wav")

    try:
        import subprocess
        # PowerShell script using System.Speech
        ps_script = f'''
        Add-Type -AssemblyName System.Speech
        $speak = New-Object System.Speech.Synthesis.SpeechSynthesizer
        $speak.SetOutputToWaveFile("{output_path}")
        $speak.Speak(''' + text.replace("'", "''") + ''')
        $speak.Dispose()
        '''
        subprocess.run(["powershell", "-Command", ps_script], check=True, capture_output=True, timeout=30)

        if os.path.exists(output_path):
            return {"audio_path": output_path, "format": "wav", "engine": "sapi"}
        else:
            return {"error": "SAPI failed to generate audio", "text": ""}
    except subprocess.TimeoutExpired:
        return {"error": "SAPI timed out", "text": ""}
    except Exception as e:
        return {"error": f"SAPI TTS failed: {str(e)}", "text": ""}


def tts(text: str, voice: str = "auto", output_path: str = None, prefer: str = "kokoro") -> dict:
    """TTS with Kokoro as primary, Windows SAPI as fallback."""
    if prefer == "kokoro":
        result = tts_kokoro(text, voice, output_path)
        if "error" not in result:
            return result
        if sys.platform == "win32":
            return tts_sapi_windows(text, output_path)
        return result
    else:
        if sys.platform == "win32":
            return tts_sapi_windows(text, output_path)
        return {"error": "No TTS engine available on this platform", "text": ""}


def main():
    parser = argparse.ArgumentParser(description="Local TTS")
    parser.add_argument("text", nargs="?", help="Text to synthesize")
    parser.add_argument("--file", help="Read text from file")
    parser.add_argument("--voice", default="af_bella",
                        help="Voice (Kokoro) or 'auto' for SAPI (default: af_bella)")
    parser.add_argument("--output", help="Output WAV file path")
    parser.add_argument("--engine", default="kokoro", choices=["kokoro", "sapi"],
                        help="TTS engine (default: kokoro, falls back to sapi)")
    args = parser.parse_args()

    if args.file:
        with open(args.file, "r", encoding="utf-8") as f:
            text = f.read().strip()
    elif args.text:
        text = args.text
    else:
        print(json.dumps({"error": "No text provided", "text": ""}))
        sys.exit(1)

    result = tts(text, args.voice, args.output, args.engine)
    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: src/main/lib/voice/tts.ts
import { execFile } from "child_process"
import { join } from "path"
import { app } from "electron"
import { readFileSync, unlinkSync } from "fs"
import type { TtsResult } from "./types"

/**
 * Synthesize text to speech using local TTS (Kokoro or Windows SAPI).
 * Returns a base64-encoded WAV audio that can be played in the renderer.
 */
export async function synthesize(
  text: string,
  options?: {
    engine?: "kokoro" | "sapi"
    voice?: string
  },
): Promise<{ audioBase64: string; format: string; error?: string }> {
  const pythonCmd = await findPython()
  if (!pythonCmd) {
    // If no Python, try Windows SAPI directly
    if (process.platform === "win32") {
      return synthesizeWithSapi(text)
    }
    return { audioBase64: "", format: "", error: "Python not found" }
  }

  try {
    const scriptPath = getScriptPath("tts.py")
    const engine = options?.engine || (await hasKokoro(pythonCmd) ? "kokoro" : "sapi")

    const args = [scriptPath, text, "--engine", engine]
    if (options?.voice) {
      args.push("--voice", options.voice)
    }

    const output = await runPython(pythonCmd, args)
    const result: TtsResult = JSON.parse(output)

    if (result.error || !result.audio_path) {
      return { audioBase64: "", format: "", error: result.error || "No audio generated" }
    }

    // Read the generated audio file as base64
    const audioBuffer = readFileSync(result.audio_path)
    const audioBase64 = audioBuffer.toString("base64")

    // Clean up temp file
    try {
      unlinkSync(result.audio_path)
    } catch {
      // Best-effort cleanup
    }

    return { audioBase64, format: result.format || "wav" }
  } catch (err) {
    const errorMsg = err instanceof Error ? err.message : String(err)
    console.error("[LocalTTS] Failed:", errorMsg)

    // Fallback: try SAPI directly
    if (process.platform === "win32") {
      return synthesizeWithSapi(text)
    }
    return { audioBase64: "", format: "", error: errorMsg }
  }
}

/**
 * Windows SAPI direct TTS (no Python needed).
 */
async function synthesizeWithSapi(text: string): Promise<{ audioBase64: string; format: string }> {
  const tempDir = app.getPath("temp")
  const outputPath = join(tempDir, `tts_sapi_${Date.now()}.wav`)

  // PowerShell script for SAPI
  const psScript = `
    Add-Type -AssemblyName System.Speech
    $speak = New-Object System.Speech.Synthesis.SpeechSynthesizer
    $speak.SetOutputToWaveFile("${outputPath.replace(/\\/g, '\\\\')}")
    $speak.Speak('${text.replace(/'/g, "''")}')
    $speak.Dispose()
  `

  return new Promise((resolve, reject) => {
    execFile("powershell", ["-Command", psScript], { timeout: 30_000 }, (error) => {
      if (error) {
        reject(new Error(`SAPI failed: ${error.message}`))
        return
      }
      try {
        const audioBuffer = readFileSync(outputPath)
        const audioBase64 = audioBuffer.toString("base64")
        try { unlinkSync(outputPath) } catch {}
        resolve({ audioBase64, format: "wav" })
      } catch (err) {
        reject(new Error(`Failed to read SAPI output: ${err}`))
      }
    })
  })
}

/**
 * Check if local TTS is available.
 */
export async function checkTtsAvailable(): Promise<boolean> {
  // Windows SAPI is always available
  if (process.platform === "win32") return true

  try {
    const pythonCmd = await findPython()
    if (!pythonCmd) return false
    return await hasKokoro(pythonCmd)
  } catch {
    return false
  }
}

async function hasKokoro(pythonCmd: string): Promise<boolean> {
  try {
    await runPython(pythonCmd, ["-c", "import kokoro; print('ok')"])
    return true
  } catch {
    return false
  }
}

async function findPython(): Promise<string | null> {
  for (const cmd of ["python3", "python"]) {
    try {
      await runCommand(cmd, ["--version"])
      return cmd
    } catch {
      continue
    }
  }
  return null
}

function getScriptPath(name: string): string {
  if (!app.isPackaged) {
    return join(app.getAppPath(), "scripts", "voice", name)
  }
  return join(process.resourcesPath, "scripts", "voice", name)
}

function runPython(cmd: string, args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, {
      timeout: 30_000,
      maxBuffer: 10 * 1024 * 1024,
      env: {
        ...process.env,
        HTTP_PROXY: "",
        HTTPS_PROXY: "",
        NO_PROXY: "*",
      },
    }, (error, stdout, stderr) => {
      if (error) {
        if (stdout.trim()) {
          try {
            const parsed = JSON.parse(stdout)
            if (parsed.audio_path) {
              resolve(stdout)
              return
            }
          } catch {}
        }
        reject(new Error(stderr.trim() || error.message))
        return
      }
      resolve(stdout.trim())
    })
  })
}

function runCommand(cmd: string, args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, { timeout: 10_000 }, (error, stdout) => {
      if (error) reject(error)
      else resolve(stdout.trim())
    })
  })
}
